From 14cf9cba3f271630e963c4f57a9a3c8cf401e79b Mon Sep 17 00:00:00 2001
From: Alex Williamson <alex.williamson@redhat.com>
Date: Thu, 2 Oct 2014 17:35:46 -0400
Subject: [pci] Add ACS quirk for Solarflare SFC9120 & SFC9140

Message-id: <20141002173545.9576.64559.stgit@gimli.home>
Patchwork-id: 96885
O-Subject: [RHEL7.1 PATCH 1/2] PCI: Add ACS quirk for Solarflare SFC9120 & SFC9140
Bugzilla: 1131552
RH-Acked-by: Don Dutile <ddutile@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Myron Stowe <mstowe@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: 1131552
Upstream: 89b51cb5c3b0e8c2b8016c60f3386a5bfa468498 (linux-next)

Solarflare confirms that these devices do not allow peer-to-peer between
functions.  Quirk them to allow IOMMU grouping to expose this isolation.

Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
Acked-by: Robert Stonehouse <rstonehouse@solarflare.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 815f724..a959855 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3570,6 +3570,21 @@ static int pci_quirk_intel_pch_acs(struct pci_dev *dev, u16 acs_flags)
  return acs_flags & ~flags ? 0 : 1;
 }
 
+static int pci_quirk_solarflare_acs(struct pci_dev *dev, u16 acs_flags)
+{
+ /*
+  * SV, TB, and UF are not relevant to multifunction endpoints.
+  *
+  * Solarflare indicates that peer-to-peer between functions is not
+  * possible, therefore RR, CR, and DT are not implemented.  Mask
+  * these out as if they were clear in the ACS capabilities register.
+  */
+ acs_flags &= ~(PCI_ACS_SV | PCI_ACS_TB | PCI_ACS_RR |
+         PCI_ACS_CR | PCI_ACS_UF | PCI_ACS_DT);
+
+ return acs_flags ? 0 : 1;
+}
+
 static const struct pci_dev_acs_enabled {
  u16 vendor;
  u16 device;
@@ -3581,6 +3596,8 @@ static const struct pci_dev_acs_enabled {
  { PCI_VENDOR_ID_ATI, 0x439d, pci_quirk_amd_sb_acs },
  { PCI_VENDOR_ID_ATI, 0x4384, pci_quirk_amd_sb_acs },
  { PCI_VENDOR_ID_ATI, 0x4399, pci_quirk_amd_sb_acs },
+ { PCI_VENDOR_ID_SOLARFLARE, 0x0903, pci_quirk_solarflare_acs },
+ { PCI_VENDOR_ID_SOLARFLARE, 0x0923, pci_quirk_solarflare_acs },
  { PCI_VENDOR_ID_INTEL, PCI_ANY_ID, pci_quirk_intel_pch_acs },
  { 0 }
 };
-- 
1.7.1